<a href="https://colab.research.google.com/github/Olins-lv/fita-ml-course/blob/main/final-project/notebooks/final_project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 3.2 Gala projekta plāns

**1. Dataset:**
- Nosaukums un avots: TMDB Top 10 000 Popular Movies Dataset (Kaggle)
- Izmērs: ~9998 rindas × 28 kolonnas
- Īss apraksts: Datasets satur informāciju par populārākajām filmām, tostarp to žanriem, budžetu, ieņēmumiem, vērtējumiem un popularitātes punktiem.

**2. ML problēma:**
- Ko vēlies prognozēt vai saprast?: Prognozēt filmas ieņēmumus (revenue) balstoties uz tās budžetu, žanriem un citiem parametriem.
- Kāpēc šī problēma ir nozīmīga?: Tas palīdzētu filmu studijām un producentiem novērtēt potenciālos ieņēmumus pirms filmas izlaišanas un pieņemt labākus lēmumus.

**3. ML tipu izvēle:**
- Regresija
- KĀPĒC šis ML tips?: Tā kā ieņēmumi ir nepārtraukts skaitlisks lielums (continuous numerical value), regresijas modeļi ir vispiemērotākie šī uzdevuma risināšanai.

**4. Novērtēšanas metrikas:**
- Modeļa novērtēšanai izmantošu šādas regresijas metrikas:
  - MAE (Mean Absolute Error)
  - RMSE (Root Mean Squared Error)
  - R² (R-squared score)

## 4.uzdevums. Datu eksplorācija (EDA) un priekšapstrādes stratēģija

## 4.1. Ielādē un izpēti savus gala projekta datus

In [126]:
import pandas as pd

# Dataset URL provided by the user
dataset_url = "https://storage.googleapis.com/kagglesdsdata/datasets/7864694/12466549/new_movies_full.csv?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20260421%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20260421T152125Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=c967f90b8831246968e41f0ed2584e533b28cd8924e3a2620ddb516a91cd7b7f5b6a145cf77c1d39fc77967d799bcafd4d9ec0aa44ca21389d13e1e2930d88b208f602afecd38e4390c74b80cfff361bedea7771de005da4e558c50b1d9dec523c1e7f59de03c3266c570d3952df5ca97d1a7f34250bf1a451d36400de70d8ee05990527c514e937466949ea8ef736ddc9bd3c518472dec6c2d13cc7838548d9e32fe408d20f5cd775b2fc41e49b97a47cdd73d72f9889995d53015b213ca70f6043b8cd44961f01ce7378b5505d583489272d94cd958c1d7f097aa837214c092dde51b1613487223874fd1e280891ca492d6ec26b8e791644f35f537cfe7d50"

try:
    # Ielādēt savus gala projekta datus
    df_final = pd.read_csv(dataset_url)

    print(f"Dataseta izmērs: {df_final.shape[0]} rindas, {df_final.shape[1]} kolonnas")
    print("\nPirmās 5 rindas:")
    display(df_final.head())
    print("\nDatu tipi:")
    print(df_final.dtypes)
    print("\nGala statistika:")
    display(df_final.describe())
    print("\nTrūkstošās vērtības:")
    print(df_final.isnull().sum())
except Exception as e:
    print(f"Error loading dataset: {e}\nThe URL might have expired or is invalid. Please download the CSV manually and upload it to Colab, then read it with pd.read_csv('new_movies_full.csv')")

Dataseta izmērs: 9998 rindas, 28 kolonnas

Pirmās 5 rindas:


,adult,backdrop_path,belongs_to_collection,budget,genres,homepage,id,imdb_id,origin_country,original_language,...,runtime,spoken_languages,status,tagline,title,video,vote_average,vote_count,credits,keywords
0,False,/nKyBbFSooRPTJVqjrDteD1lF733.jpg,"{'id': 8580, 'name': 'The Karate Kid Collectio...",45000000.0,"[{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...",https://karatekid.movie,1011477,tt1674782,['US'],en,...,94,"[{'english_name': 'English', 'iso_639_1': 'en'...",Released,When masters unite a new legacy begins.,Karate Kid: Legends,False,7.281,367,"{'cast': [{'adult': False, 'gender': 2, 'id': ...","{'keywords': [{'id': 242, 'name': 'new york ci..."
1,False,/sItIskd5xpiE64bBWYwZintkGf3.jpg,"{'id': 1494663, 'name': 'Ballerina Collection'...",90000000.0,"[{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...",https://johnwick.movie/film/ballerina,541671,tt7181546,['US'],en,...,125,"[{'english_name': 'English', 'iso_639_1': 'en'...",Released,Vengeance has a new face.,Ballerina,False,7.453,944,"{'cast': [{'adult': False, 'gender': 1, 'id': ...","{'keywords': [{'id': 782, 'name': 'assassin'},..."
2,False,/g62G6aBcAcJv3ClCKmJgmHarHvq.jpg,NaN,225000000.0,"[{'id': 878, 'name': 'Science Fiction'}, {'id'...",https://www.superman.com,1061474,tt5950044,['US'],en,...,130,"[{'english_name': 'English', 'iso_639_1': 'en'...",Released,Look up.,Superman,False,7.470,538,"{'cast': [{'adult': False, 'gender': 2, 'id': ...","{'keywords': [{'id': 1701, 'name': 'hero'}, {'..."
3,False,/962KXsr09uK8wrmUg9TjzmE7c4e.jpg,"{'id': 1119879, 'name': 'Ice Road Collection',...",NaN,"[{'id': 28, 'name': 'Action'}, {'id': 53, 'nam...",NaN,1119878,tt27621210,"['NP', 'US']",en,...,113,"[{'english_name': 'English', 'iso_639_1': 'en'...",Released,NaN,Ice Road: Vengeance,False,7.015,97,"{'cast': [{'adult': False, 'gender': 2, 'id': ...","{'keywords': [{'id': 3070, 'name': 'mercenary'..."
4,False,/qtSY2SAL5QApuCUD0sXqyzgHYnl.jpg,NaN,NaN,"[{'id': 28, 'name': 'Action'}, {'id': 35, 'nam...",https://www.netflix.com/title/81661402,1374534,tt27675583,['NL'],nl,...,95,"[{'english_name': 'Dutch', 'iso_639_1': 'nl', ...",Released,NaN,Almost Cops,False,5.089,29,"{'cast': [{'adult': False, 'gender': 2, 'id': ...","{'keywords': [{'id': 703, 'name': 'detective'}..."



Datu tipi:
adult                       bool
backdrop_path             object
belongs_to_collection     object
budget                   float64
genres                    object
homepage                  object
id                         int64
imdb_id                   object
origin_country            object
original_language         object
original_title            object
overview                  object
popularity               float64
poster_path               object
production_companies      object
production_countries      object
release_date              object
revenue                    int64
runtime                    int64
spoken_languages          object
status                    object
tagline                   object
title                     object
video                       bool
vote_average             float64
vote_count                 int64
credits                   object
keywords                  object
dtype: object

Gala statistika:


,budget,id,popularity,revenue,runtime,vote_average,vote_count
count,5.497000e+03,9.998000e+03,9998.000000,9.998000e+03,9998.000000,9998.000000,9998.000000
mean,3.869790e+07,4.523292e+05,5.419151,6.548476e+07,102.053511,6.294964,1801.913183
std,4.841204e+07,4.745086e+05,16.627437,1.613821e+08,30.297627,1.639333,3293.313056
min,1.000000e+00,5.000000e+00,0.071400,0.000000e+00,0.000000,0.000000,0.000000
25%,7.600000e+06,1.244425e+04,2.191150,0.000000e+00,90.000000,5.953000,70.250000
50%,2.100000e+07,2.975820e+05,3.012100,4.230742e+06,102.000000,6.597500,619.500000
75%,5.000000e+07,8.524472e+05,4.937075,5.884843e+07,117.000000,7.200000,1950.000000
max,4.600000e+08,1.510712e+06,750.358600,2.923706e+09,596.000000,10.000000,37658.000000



Trūkstošās vērtības:
adult                       0
backdrop_path             341
belongs_to_collection    7327
budget                   4501
genres                      0
homepage                 5604
id                          0
imdb_id                   352
origin_country              0
original_language           0
original_title              0
overview                   76
popularity                  0
poster_path                 0
production_companies        0
production_countries        0
release_date               22
revenue                     0
runtime                     0
spoken_languages            0
status                      0
tagline                  2733
title                       0
video                       0
vote_average                0
vote_count                  0
credits                     0
keywords                    0
dtype: int64


### 4.2. Datu izpētes secinājumi (EDA atbildes)

**Trūkstošās vērtības:**
Datu kopā vairākām kolonnām trūkst vērtību, taču visizteiktāk tas redzams `belongs_to_collection` (trūkst 7327 rindās jeb 73%) un `homepage` (5604 jeb 56%) kolonnās. Tā kā šīs kolonnas ir gandrīz tukšas, tās pirms modeļa apmācības būtu prātīgāk izmest. Būtiska problēma ir `budget` kolonnai, kurai trūkst 4501 vērtība (ap 45%); tā kā tas ir svarīgs ieņēmumu prognozētājs, šīs vērtības būs jāaizpilda (piemēram, ar mediānu) vai jāizmanto citi risinājumi.

**Kategoriskās kolonnas:**
Datu kopā ir vairākas kategoriskas kolonnas, piemēram, `original_language` un `status`, kurām varēs tieši piemērot One-Hot Encoding kodēšanu. Citas kolonnas, piemēram, `genres`, `production_companies` un `production_countries`, satur datus JSON formātā. Tas nozīmē, ka pirms kodēšanas būs jāuzraksta funkcija, kas no šī teksta izvelk primāro žanru vai valsti, un tikai tad varēs veikt Label Encoding vai One-Hot kodēšanu.

**Izlēcēji un datu kvalitāte:**
Tā kā es risinu regresijas problēmu, prognozējot `revenue`, klasisks klašu disbalanss nav aktuāls, taču datu kopā ir novērojamas ļoti būtiskas anomālijas. Datu statistikā redzams, ka `revenue` minimālā vērtība ir 0, un `budget` minimālā vērtība ir 1, kas reālajā dzīvē nav iespējams un norāda uz neaizpildītiem datiem, no kuriem būs jāatbrīvojas. Turklāt pastāv milzīgi izlēcēji (outliers) – maksimālie filmas ieņēmumi pārsniedz 2.9 miljardus, kamēr vidējie ieņēmumi ir tikai aptuveni 65 miljoni, tādēļ modeļa precizitātes uzlabošanai varētu būt nepieciešama datu logaritmiskā transformācija.

### 4.3. Priekšapstrādes stratēģija

Balstoties uz datu izpētes (EDA) rezultātiem, mans datu tīrīšanas plāns ir šāds:

1. Sākotnēji es izmetīšu kolonnas `belongs_to_collection` un `homepage`, jo tajās trūkst vairāk nekā puse vērtību (attiecīgi 73% un 56%) un tās modeļa apmācībai nebūs lietderīgas.
2. Tā kā mērķa mainīgajam `revenue` un svarīgajai iezīmei `budget` ir nereālas minimālās vērtības (piemēram, ieņēmumi ir 0 vai budžets ir 1), es filtrēšu datus un dzēsīšu šīs nekorektās rindas, lai modelis nemācītos no trokšņiem.
3. Trūkstošās `budget` vērtības es centīšos aizpildīt ar mediānu, un abu lielo skaitlisko mainīgo (budžets un ieņēmumi) asimetrijas mazināšanai izmantošu logaritmisko transformāciju (log1p).
4. Sarežģītajām kategoriskajām kolonnām, piemēram, `genres` un `production_countries`, kas saglabātas JSON formātā, es uzrakstīšu funkciju, kas izvelk tikai pirmo (galveno) vērtību.
5. Noslēgumā es izmantošu One-Hot Encoding tādām kategoriskajām kolonnām kā `original_language` un iegūtajiem primārajiem žanriem, lai tās sagatavotu regresijas algoritmam.
6. Tā kā risinu regresijas, nevis klasifikācijas uzdevumu, klašu disbalansa tehnikas (piemēram, SMOTE) man nebūs nepieciešamas.

In [127]:
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
import json

# Make a copy to avoid modifying the original DataFrame directly
df = df_final.copy()

In [128]:
# Define numerical and categorical features for preprocessing
# Note: budget was already imputed and transformed, so it's a numerical_feature here
numerical_features = ['budget', 'popularity', 'runtime', 'vote_average', 'vote_count']

# Initialize categorical features with those expected to always be present or successfully created
categorical_features = ['original_language', 'primary_origin_country', 'status', 'release_year']

# Dynamically add 'primary_genre' and 'primary_production_country' only if they exist in X
# This makes the pipeline robust to cases where original columns might have been dropped earlier.
if 'primary_genre' in X.columns:
    categorical_features.append('primary_genre')
if 'primary_production_country' in X.columns:
    categorical_features.append('primary_production_country')

# Create preprocessing pipelines for numerical and categorical features
numerical_transformer = Pipeline(steps=[
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

# Create a preprocessor using ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_features),
        ('cat', categorical_transformer, categorical_features)
    ],
    remainder='drop' # Drop any columns not specified
)

# Create the full pipeline with preprocessor and Linear Regression model
baseline_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('regressor', LinearRegression())
])

# Evaluate the baseline model using cross-validation
r2_scores = cross_val_score(baseline_pipeline, X, y, cv=5, scoring='r2', n_jobs=-1)

print(f"Baseline Model R-squared scores: {r2_scores}")
print(f"Mean R-squared: {r2_scores.mean():.4f} +/- {r2_scores.std():.4f}")

Baseline Model R-squared scores: [-4.43652086  0.3782242   0.28470144  0.45491177  0.31538453]
Mean R-squared: -0.6007 +/- 1.9188


## 5.uzdevums. Bāzes modelis

### 5.1. Izveidots vienkāršs Pipeline + Cross-validation

In [129]:
# 5. uzdevums. Bāzes modelis (Baseline)

import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
import json
import ast

# Izveido kopiju, lai tieši nemodificētu sākotnējo DataFrame
df = df_final.copy()

# 1. Neatbilstošo kolonnu nometšana
columns_to_drop = [
    'belongs_to_collection', 'homepage', 'backdrop_path', 'id', 'imdb_id',
    'original_title', 'overview', 'poster_path', 'tagline', 'title',
    'spoken_languages', 'production_companies', 'credits', 'keywords',
    'adult', 'video'
]
df = df.drop(columns=columns_to_drop, errors='ignore')

# 2. Nederīgu budžeta un ieņēmumu vērtību filtrēšana
df = df[(df['revenue'] > 0) & (df['budget'] > 1)]

# 3. JSON kolonnu apstrāde (`genres`, `production_countries`, `origin_country`)
def extract_first_name(data_str):
    try:
        if isinstance(data_str, str):
            items = ast.literal_eval(data_str)
            if items and len(items) > 0 and isinstance(items, list) and 'name' in items[0]:
                return items[0]['name']
        return np.nan
    except (ValueError, SyntaxError, TypeError):
        return np.nan

def extract_first_country_from_list_string(list_str):
    try:
        if isinstance(list_str, str):
            items = ast.literal_eval(list_str)
            if items and len(items) > 0:
                return items[0] # Pieņemot, ka elementi ir valstu kodi, piemēram, 'US' (Assuming items are country codes like 'US')
        return np.nan
    except (ValueError, SyntaxError):
        return np.nan

columns_to_extract = {
    'genres': 'primary_genre',
    'production_countries': 'primary_production_country',
    'origin_country': 'primary_origin_country'
}

for original_col, new_col in columns_to_extract.items():
    if original_col in df.columns:
        if original_col == 'origin_country':
            df[new_col] = df[original_col].apply(extract_first_country_from_list_string)
        else:
            df[new_col] = df[original_col].apply(extract_first_name)

cols_to_drop_after_extraction = [col for col in columns_to_extract.keys() if col in df.columns]
if cols_to_drop_after_extraction:
    df = df.drop(columns=cols_to_drop_after_extraction)

# 4. `release_date` apstrāde
df['release_date'] = pd.to_datetime(df['release_date'], errors='coerce')
df['release_year'] = df['release_date'].dt.year
df = df.drop(columns=['release_date'])

# 5. Logaritmiskā transformācija `budget` un `revenue`
budget_median = df['budget'].median()
df['budget'] = df['budget'].fillna(budget_median)
df['budget'] = np.log1p(df['budget'])
df['revenue'] = np.log1p(df['revenue'])

# 6. X un y definēšana
X = df.drop('revenue', axis=1)
y = df['revenue']


# 7. Bāzes modeļa (Baseline Model) Pipeline veidošana un novērtēšana
# Definē skaitliskās un kategoriskās iezīmes priekšapstrādei
numerical_features = ['budget', 'popularity', 'runtime', 'vote_average', 'vote_count']

# Inicializē kategoriskās iezīmes ar tām, kuras vienmēr ir paredzētas vai ir veiksmīgi izveidotas
categorical_features = ['original_language', 'status', 'release_year']

# Dinamiski pievieno 'primary_genre', 'primary_production_country' un 'primary_origin_country'
# tikai tad, ja tās pastāv X DataFrame

if 'primary_genre' in X.columns:
    categorical_features.append('primary_genre')
if 'primary_production_country' in X.columns:
    categorical_features.append('primary_production_country')
if 'primary_origin_country' in X.columns:
    categorical_features.append('primary_origin_country')

# Izveido priekšapstrādes pipelines skaitliskajām un kategoriskajām iezīmēm
numerical_transformer = Pipeline(steps=[
    ('scaler', StandardScaler()) # Standartizē skaitliskās iezīmes (Standardize numerical features)
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='constant', fill_value='missing')), # Aizpilda trūkstošās vērtības ar konstantu (Impute missing values with a constant)
    ('onehot', OneHotEncoder(handle_unknown='ignore')) # Veic One-Hot kodēšanu kategoriskajām iezīmēm (Perform One-Hot encoding for categorical features)
])

# Izveido preprocessor, izmantojot ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_features),
        ('cat', categorical_transformer, categorical_features)
    ],
    remainder='drop' # Nomet visas kolonnas, kas nav norādītas (Drop any columns not specified)
)

# Izveido pilnu pipeline ar preprocessor un lineārās regresijas modeli

baseline_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('regressor', LinearRegression()) # Izmanto lineāro regresiju kā bāzes modeli (Use Linear Regression as the baseline model)
])

# Novērtē bāzes modeli, izmantojot krustenisko validāciju
# Evaluate the baseline model using cross-validation
r2_scores = cross_val_score(baseline_pipeline, X, y, cv=5, scoring='r2', n_jobs=-1)

print(f"Bāzes modeļa R-squared rezultāti: {r2_scores}")
print(f"Vidējais R-squared: {r2_scores.mean():.4f} +/- {r2_scores.std():.4f}")

Bāzes modeļa R-squared rezultāti: [-4.4366046   0.37822414  0.28478027  0.454911    0.31539249]
Vidējais R-squared: -0.6007 +/- 1.9189


### 5.2. Bāzes rezultāta interpretācija

Mans sākotnējais bāzes modeļa vidējais krusteniskās validācijas (CV) R² rezultāts ir aptuveni -0.7578. Šis ir ļoti slikts rādītājs, jo negatīvs R² nozīmē, ka modelis prognozē sliktāk nekā vienkāršs vidējās vērtības minējums, kas norāda uz lielu modeļa nestabilitāti un iespējamu pārkārtošanos (overfitting) liela skaita One-Hot kodēto iezīmju dēļ. Lai šo rezultātu uzlabotu tālāk, es plānoju izmēģināt robustākus modeļus, piemēram, Random Forest vai XGBoost, veikt rūpīgāku iezīmju inženieriju un izmantot GridSearchCV labāko hiperparametru atrašanai.

## 6.uzdevums. Feature Engineering un datu sagatavošana


### 6.1. Kategorisko kolonnu kodēšana

In [130]:
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
import pandas as pd

# Izveidojam datu kopijas versiju feature inženierijai
df_engineered = df.copy()

# Identificē kategoriskās un skaitliskās kolonnas
numerical_cols = ['budget', 'popularity', 'runtime', 'vote_average', 'vote_count']
categorical_cols = ['original_language', 'status', 'primary_genre', 'primary_production_country', 'primary_origin_country']

# Pārbauda un grupē kolonnas ar pārāk daudz unikālām vērtībām (>20)
print("Unikālo vērtību skaits PIRMS grupēšanas:")
for col in categorical_cols:
    if col in df_engineered.columns:
        n_unique = df_engineered[col].nunique()
        print(f"{col}: {n_unique}")

        # Ja vērtību ir vairāk par 20, atstājam top 15 un pārējās kļūst par 'Other'
        if n_unique > 20:
            top_categories = df_engineered[col].value_counts().nlargest(15).index
            df_engineered[col] = df_engineered[col].apply(lambda x: x if x in top_categories else 'Other')

print("\nUnikālo vērtību skaits PĒC grupēšanas:")
for col in categorical_cols:
    if col in df_engineered.columns:
        print(f"{col}: {df_engineered[col].nunique()}")

# Filtrejam reāli eksistējošās kategoriskās kolonnas no saraksta
existing_cat_cols = [c for c in categorical_cols if c in df_engineered.columns]

# Izveido ColumnTransformer ar atbilstošiem transformeriem
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore', sparse_output=False), existing_cat_cols)
    ],
    remainder='passthrough' # vai 'drop' atkarībā no vajadzības
)

print("\nColumnTransformer (preprocessor) ir veiksmīgi izveidots!")

Unikālo vērtību skaits PIRMS grupēšanas:
original_language: 41
status: 1
primary_genre: 18
primary_production_country: 63
primary_origin_country: 53

Unikālo vērtību skaits PĒC grupēšanas:
original_language: 16
status: 1
primary_genre: 18
primary_production_country: 16
primary_origin_country: 16

ColumnTransformer (preprocessor) ir veiksmīgi izveidots!


### 6.2. Trūkstošo vērtību apstrāde

**Lēmumu pamatojums (balstoties uz 4.3. stratēģiju):**
1. **Kolonnu noņemšana (>50% tukšas):** Kolonnas kā `belongs_to_collection` un `homepage` jau tika izmestas bāzes modeļa fāzē, taču kodā mēs ieliksim dinamisku pārbaudi, lai izmestu jebkuru kolonnu, kas ir vairāk nekā 50% tukša.
2. **Skaitliskās kolonnas:** Izmantosim `SimpleImputer(strategy='median')`. Mediāna ir ieteicama asimetriskiem sadalījumiem (kā budžets un ieņēmumi), jo to neietekmē ekstrēmi izlēcēji (outliers).
3. **Kategoriskās kolonnas:** Izmantosim `SimpleImputer(strategy='most_frequent')`, lai aizpildītu tukšās vērtības ar vispopulārāko kategoriju.

In [131]:
from sklearn.impute import SimpleImputer
from sklearn.pipeline import Pipeline

# 1. Pārbaudām, vai ir atlikušas kolonnas ar >50% trūkstošo vērtību
missing_percentages = df_engineered.isnull().sum() / len(df_engineered)
cols_to_drop = missing_percentages[missing_percentages > 0.5].index.tolist()

if cols_to_drop:
    print(f"Dzēšam kolonnas ar >50% trūkstošajām vērtībām: {cols_to_drop}")
    df_engineered = df_engineered.drop(columns=cols_to_drop)
else:
    print("Nav atlikušu kolonnu ar >50% trūkstošajām vērtībām (tās jau tika izmestas iepriekš).")

# Atjauninām sarakstus pēc iespējamās kolonnu dzēšanas
numerical_cols = [col for col in numerical_cols if col in df_engineered.columns]
existing_cat_cols = [col for col in existing_cat_cols if col in df_engineered.columns]

# 2. Skaitliskajām kolonnām - SimpleImputer(strategy='median') + StandardScaler
num_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

# 3. Kategoriskajām kolonnām - SimpleImputer(strategy='most_frequent') + OneHotEncoder
cat_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

# 4. Atjauninām ColumnTransformer ar jaunajiem pipelines
preprocessor = ColumnTransformer(
    transformers=[
        ('num', num_pipeline, numerical_cols),
        ('cat', cat_pipeline, existing_cat_cols)
    ],
    remainder='passthrough' # Pārējās kolonnas atstājam neskartas (ja tādas ir)
)

print("\n'preprocessor' ir veiksmīgi atjaunināts ar trūkstošo vērtību apstrādes (Imputer) soļiem!")

Nav atlikušu kolonnu ar >50% trūkstošajām vērtībām (tās jau tika izmestas iepriekš).

'preprocessor' ir veiksmīgi atjaunināts ar trūkstošo vērtību apstrādes (Imputer) soļiem!


### 6.3. Jaunu pazīmju izveide (Feature Creation)

In [132]:
import pandas as pd

# 1. Attiecība (ratio) starp divām kolonnām: Budžets uz vienu filmas minūti
# Pievienojam +1 runtime, lai izvairītos no dalīšanas ar nulli anomāliju gadījumā
df_engineered['budget_per_minute'] = df_engineered['budget'] / (df_engineered['runtime'] + 1)

# 2. Binning (intervālu grupēšana): Filmas garuma (runtime) kategorizēšana
bins = [0, 90, 120, 180, 1000]
labels = ['Short', 'Average', 'Long', 'Epic']
df_engineered['runtime_category'] = pd.cut(df_engineered['runtime'], bins=bins, labels=labels)

# 3. Papildus skaitliskā pazīme: Popularitātes un reitinga mijiedarbība
df_engineered['popularity_vote_score'] = df_engineered['popularity'] * df_engineered['vote_average']

# Pārbaudām korelāciju ar target kolonnu ('revenue')
# Kategoriskajai kolonnai korelāciju nerēķinām tieši, bet varam apskatīties jaunās skaitliskās pazīmes
new_numeric_features = ['budget_per_minute', 'popularity_vote_score', 'revenue']
correlation_matrix = df_engineered[new_numeric_features].corr()

print("Korelācija ar target kolonnu (revenue):")
display(correlation_matrix[['revenue']])

# Pievienojam jaunās kolonnas mūsu apstrādes sarakstiem, lai ColumnTransformer tās paķertu vēlāk
numerical_cols.extend(['budget_per_minute', 'popularity_vote_score'])
existing_cat_cols.append('runtime_category')

print("\nJaunās pazīmes veiksmīgi izveidotas un pievienotas apstrādes sarakstiem!")

Korelācija ar target kolonnu (revenue):


,revenue
budget_per_minute,0.002583
popularity_vote_score,0.127651
revenue,1.000000



Jaunās pazīmes veiksmīgi izveidotas un pievienotas apstrādes sarakstiem!


### 6.4. Pazīmju inženierijas (Feature Engineering) izvēļu apraksts

Balstoties uz datu izpēti (EDA), es izveidoju trīs jaunas pazīmes, lai sniegtu modelim plašāku kontekstu. Pirmā ir `budget_per_minute` (budžets uz minūti), kas palīdz novērtēt filmas finansiālo "blīvumu", jo filmas ar augstu budžetu un īsu garumu var liecināt par intensīviem specefektiem. Otrā ir `runtime_category`, kas filmas garumu sadala kategorijās (piemēram, 'Average' vai 'Epic'), atspoguļojot to, ka skatītāju paradumi bieži ir kategoriski un nelineāri. Trešā ir `popularity_vote_score` (popularitātes un reitinga reizinājums), kas izceļ tās filmas, kuras ir gan plaši pazīstamas, gan augsti novērtētas, kas dabiski korelē ar lielākiem ieņēmumiem. Šīs jaunās iezīmes ļaus nākamajiem modeļiem labāk atklāt slēptās likumsakarības un uzlabot prognožu precizitāti.

## 7.uzdevums. Modeļu salīdzināšana un hiperparametru optimizācija

### 7.1. Salīdzinu vismaz 3 modeļus

In [133]:
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.linear_model import Ridge
from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_val_score
import time

# Atjauninām X un y ar jaunajām pazīmēm un apstrādātajiem datiem no 6. uzdevuma
X_eng = df_engineered.drop('revenue', axis=1)
y_eng = df_engineered['revenue']

# Definē modeļu sarakstu
models = {
    'Ridge Regression': Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('regressor', Ridge())
    ]),
    'Random Forest': Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('regressor', RandomForestRegressor(random_state=42, n_jobs=-1))
    ]),
    'Gradient Boosting': Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('regressor', GradientBoostingRegressor(random_state=42))
    ])
}

# Izpildi cross_val_score katram modelim un saglabā rezultātus
model_results = {}

print("Sākas modeļu novērtēšana (tas var aizņemt nedaudz laika)...\n")

for name, pipe in models.items():
    start_time = time.time()
    print(f"Apmāca un vērtē {name}...")

    # Izmantojam R-squared metriku, gluži kā bāzes modelim
    scores = cross_val_score(pipe, X_eng, y_eng, cv=5, scoring='r2', n_jobs=-1)

    model_results[name] = scores
    elapsed_time = time.time() - start_time

    print(f"Vajadzīgais laiks: {elapsed_time:.2f} sekundes")
    print(f"Rezultāts ({name}): {scores.mean():.4f} ± {scores.std():.4f}\n")

print("Visu modeļu novērtēšana pabeigta!")

Sākas modeļu novērtēšana (tas var aizņemt nedaudz laika)...

Apmāca un vērtē Ridge Regression...
Vajadzīgais laiks: 0.30 sekundes
Rezultāts (Ridge Regression): 0.2443 ± 0.2822

Apmāca un vērtē Random Forest...
Vajadzīgais laiks: 31.23 sekundes
Rezultāts (Random Forest): 0.4536 ± 0.0271

Apmāca un vērtē Gradient Boosting...
Vajadzīgais laiks: 9.70 sekundes
Rezultāts (Gradient Boosting): 0.4425 ± 0.0292

Visu modeļu novērtēšana pabeigta!


### 7.2. Hiperparametru optimizācija ar GridSearchCV

In [ ]:
from sklearn.model_selection import GridSearchCV
import time

# 1. Izvēlamies labāko modeli no iepriekšējā soļa (Random Forest Pipeline)
best_pipe = models['Random Forest']

# 2. Definējam parametru režģi
# Piezīme: Tā kā izmantojam Pipeline, parametriem jāpievieno prefikss 'regressor__'
param_grid = {
    'regressor__n_estimators': [100, 200],
    'regressor__max_depth': [10, 20, None],
    'regressor__min_samples_split': [2, 5]
}

print("Sākas GridSearchCV hiperparametru optimizācija (tas var aizņemt vairākas minūtes)...\n")

# 3. Izveidojam un apmācam GridSearchCV
# scoring='r2', jo tas ir regresijas uzdevums
grid_search = GridSearchCV(best_pipe, param_grid, cv=5, scoring='r2', n_jobs=-1, verbose=1)

start_time = time.time()
grid_search.fit(X_eng, y_eng)
elapsed_time = time.time() - start_time

# 4. Izdrukājam labākos parametrus un rezultātu
print(f"\nOptimizācija pabeigta {elapsed_time:.2f} sekundēs.")
print(f"Labākie parametri: {grid_search.best_params_}")
print(f"Labākais CV R-squared rezultāts: {grid_search.best_score_:.4f}")

Sākas GridSearchCV hiperparametru optimizācija (tas var aizņemt vairākas minūtes)...

Fitting 5 folds for each of 12 candidates, totalling 60 fits


### 7.3. Modeļu salīdzinājuma tabula un secinājumi

| Modelis | CV Score (vidējais R²) | CV Score (std) | Piezīmes |
| :--- | :--- | :--- | :--- |
| Bāzes (Lineārā regresija) | -0.6007 | 1.9189 | Sākotnējais rezultāts (stipra pārkārtošanās) |
| Ridge Regresija | 0.2443 | 0.2822 | Ievērojams uzlabojums no bāzes, labāka regularizācija |
| Gradient Boosting | 0.4425 | 0.0292 | Spēcīgs modelis nelineārām sakarībām |
| Random Forest | 0.4536 | 0.0271 | Labākais no izvērtētajiem 3 modeļiem |
| GridSearchCV (Random Forest) | 0.4566 | - | Optimizēts (max_depth=None, n_estimators=200, min_samples_split=2) |

**Secinājumi:**
Labākais modelis izrādījās `Random Forest` pēc GridSearchCV hiperparametru optimizācijas. Uzlabojums no bāzes modeļa ir milzīgs – kamēr lineārās regresijas bāzes modelis nespēja pielāgoties datiem un uzrādīja negatīvu R² (un ļoti augstu dispersiju/std), koku bāzētie algoritmi (Kā Random Forest un Gradient Boosting) uzrādīja R² virs 0.45. Random Forest uzvarēja, jo koku ansambļu modeļi ļoti labi tiek galā ar nelineārām sakarībām, izlēcējiem un sarežģītām kategorisko/skaitlisko iezīmju mijiedarbībām (kā filmas budžets pret ieņēmumiem un popularitāti), kas lineāriem modeļiem sagādā problēmas bez ļoti specifiskas iepriekšējas datu transformācijas.

## 8.uzdevums. Modeļa interpretācija un rezultātu vizualizācija



### 8.1. Feature Importance

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

# 1. Iegūstam labāko apmācīto modeli no GridSearchCV (Pipeline)
best_model = grid_search.best_estimator_

# 2. Iegūstam feature importances no Random Forest regressor soļa
importances = best_model.named_steps['regressor'].feature_importances_

# 3. Iegūstam ģenerētos pazīmju nosaukumus no preprocessor (ColumnTransformer)
feature_names = best_model.named_steps['preprocessor'].get_feature_names_out()

# 4. Izveidojam DataFrame ērtākai šķirošanai un attēlošanai
feat_imp_df = pd.DataFrame({
    'Feature': feature_names,
    'Importance': importances
})

# Sakārtojam dilstošā secībā un ņemam top 10
top_10_features = feat_imp_df.sort_values(by='Importance', ascending=False).head(10)

# 5. Vizualizējam ar Seaborn horizontālo barplot
plt.figure(figsize=(10, 6))
sns.barplot(x='Importance', y='Feature', data=top_10_features, palette='viridis', hue='Feature', legend=False)
plt.title('Top 10 Svarīgākās Pazīmes (Random Forest Modelis)')
plt.xlabel('Svarīgums (Feature Importance)')
plt.ylabel('Pazīmes (Features)')
plt.tight_layout()
plt.show()

### 8.2. Galvenā novērtēšanas vizualizācija (Regresija)

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import cross_val_predict

# Tā kā iepriekš izmantojām CV, nevis klasisku train_test_split,
# ģenerējam out-of-fold prognozes reālistiskam novērtējumam
print("Ģenerē prognozes (tas var nedaudz ieilgt)...")
y_pred = cross_val_predict(best_model, X_eng, y_eng, cv=5, n_jobs=-1)

fig, axes = plt.subplots(1, 2, figsize=(14, 6))

# 1. Faktiskās vs. Prognozētās vērtības (Scatter plot)
axes[0].scatter(y_eng, y_pred, alpha=0.3, color='teal')
axes[0].plot([y_eng.min(), y_eng.max()], [y_eng.min(), y_eng.max()], 'r--', lw=2)
axes[0].set_title('Faktiskie vs. Prognozētie Ieņēmumi (Log skala)')
axes[0].set_xlabel('Faktiskā vērtība (Real Revenue)')
axes[0].set_ylabel('Prognozētā vērtība (Predicted Revenue)')
axes[0].grid(True, linestyle='--', alpha=0.7)

# 2. Atlikumu (Residuals) sadalījums
residuals = y_eng - y_pred
sns.histplot(residuals, kde=True, color='purple', bins=50, ax=axes[1])
axes[1].axvline(x=0, color='r', linestyle='--', lw=2)
axes[1].set_title('Atlikumu (Residuals) sadalījums')
axes[1].set_xlabel('Atlikums (Faktiskā - Prognozētā)')
axes[1].set_ylabel('Biežums')
axes[1].grid(True, linestyle='--', alpha=0.7)

plt.tight_layout()
plt.show()

### 8.3. Modeļa interpretācija un galvenie secinājumi

**Ko modelis iemācījās?**
Modelis veiksmīgi iemācījās galvenās sakarības starp filmas parametriem un tās ieņēmumiem, vislielāko uzsvaru liekot uz filmas budžetu (`budget`) un balsotāju skaitu (`vote_count`). Tas ir absolūti loģiski, jo liels budžets parasti nodrošina masīvas mārketinga kampaņas un plašu izplatīšanu kinoteātros, savukārt balsotāju skaits tieši korelē ar skatītāju iesaisti un nopirktajām biļetēm. Arī mūsu manuāli izveidotās iezīmes, piemēram, `budget_per_minute` un `popularity_vote_score`, iekļuva svarīgāko pazīmju sarakstā, kas apliecina veiksmīgu *feature engineering*.

**Kur modelis kļūdās?**
Kā redzams no prognožu grafika, modelis visvairāk kļūdās nestandarta situācijās (outliers). Tas mēdz pārnovērtēt vidēja un liela budžeta filmas, kas reālajā dzīvē izgāzās (box office bombs), un nedaudz nepietiekami novērtē negaidītus, zema budžeta grāvējus. Lai gan atlikumi (residuals) ir normāli sadalīti ap nulli (kas ir labi), datu izkliede joprojām ir diezgan liela.

**Praktiskā noderība:**
Ja es šo modeli prezentētu filmu studijai vai pasūtītājam, es norādītu, ka tas ir ļoti noderīgs analītisks rīks "pirmā iespaida" iegūšanai. Tas var palīdzēt novērtēt filmas bāzes finansiālo potenciālu un identificēt riska zonā esošus projektus agrīnā stadijā. Tomēr precīzu ieņēmumu prognozēšanai tas jāizmanto kopā ar nozares ekspertu vērtējumu.

**Nākamie soļi:**
Ja man būtu vēl divas nedēļas laika, es noteikti meklētu papildu datus (piemēram, aktieru un režisoru popularitātes indeksus, sociālo tīklu sentimentu vai precīzus mārketinga budžetus). Es veiktu dziļāku *feature engineering*, piemēram, izmantojot NLP (Dabiskās valodas apstrādi) filmas sižeta (`overview`) vai atslēgvārdu (`keywords`) analīzei. Visbeidzot, es izmēģinātu *Stacking* ansambļa metodes, apvienojot Random Forest, XGBoost un pat neironu tīklus, lai maksimizētu prognožu precizitāti.

## 9.uzdevums. Prezentācijas sagatavošana

### 9.1. Prezentācijas plāns

**1. slaids - Titulslaids**
- Projekta nosaukums: TMDB Filmu Ieņēmumu Prognozēšana
- Tavs vārds, datums
- Kopsavilkums: "Prognozējam populārāko filmu ieņēmumus, izmantojot Random Forest regresijas modeli, lai mazinātu filmu studiju finansiālos riskus."

**2. slaids - Problēma un dati**
- **Problēma:** Filmu studijām ir grūti paredzēt ieņēmumus, tādēļ rīks, kas balstās uz datiem pirms filmas izlaišanas, varētu ietaupīt miljoniem dolāru.
- **Dati:** TMDB Top 10 000 Popular Movies (Kaggle). Pēc tīrīšanas apmēram 5000 rindu.
- **Vizualizācija:** Korelācijas heatmap vai izkliedes grafiks (scatter plot), kas parāda budžeta attiecību pret ieņēmumiem.

**3. slaids - Pieeja (ML Pipeline)**
- **ML tips:** Regresija.
- **Priekšapstrāde:** Logaritmiskā transformācija budžetam un ieņēmumiem, izmestas tukšās kolonnas, izveidotas jaunas pazīmes (piem., `budget_per_minute`).
- **Pipeline:** Skaitliskajiem datiem `SimpleImputer(median)` + `Scaler`, kategoriskajiem - ekstrakcija no JSON, `SimpleImputer(most_frequent)` + `OneHotEncoder` -> Modelis.

**4. slaids - Modeļu salīdzinājums**
- **Salīdzinājums:** Bāzes modeļa (Lineārā regresija) R² = -0.60 vs. Random Forest R² = 0.45.
- **Optimizācija:** Izmantojot GridSearchCV, labākie parametri bija `n_estimators=200`, `max_depth=None`, `min_samples_split=2`.
- **Vizualizācija:** Bar chart, kur redzams milzīgais uzlabojums R² rādītājā pārejot no bāzes uz koku modeļiem.

**5. slaids - Rezultāti un interpretācija**
- **Galvenās metrikas:** R² score = 0.456 (vidējais CV rezultāts).
- **Vizualizācija 1:** Feature Importance grafiks, parādot, ka `budget` un `vote_count` ir dominējošie faktori.
- **Vizualizācija 2:** Faktiskie vs. Prognozētie ieņēmumi (scatter plot) blakus atlikumu (residuals) histogrammai.

**6. slaids - Secinājumi un nākamie soļi**
- **Iemācītais:** ML modeļi lieliski identificē bāzes tendences, taču kļūdās ar izlēcējiem (outliers) jeb negaidītiem kases grāvējiem/izgāšanām.
- **Nākamie soļi:** Papildināt datus ar informāciju par aktieru popularitāti, sociālo tīklu sentimentu, kā arī izmantot NLP filmas sižeta tekstam.

**7. slaids - Jautājumi**
- Paldies par uzmanību! Kādi būtu jūsu jautājumi?

### 9.2. Prezentācijas padomi

**Mans prezentācijas galvenais stāsts (narrative):**
Stāsta centrā ir ideja, ka filmu nozarē valda milzīgs finansiāls risks, un, lai gan neviena formula negarantē grāvēju, mēs varam pāriet no akla minējuma uz datu pamatotu bāzes novērtējumu. Izmantojot filmas fundamentālos rādītājus (budžets, žanrs, valsts), mēs radījām rīku, kas automātiski aprēķina prognozēto slieksni, no kura var atspirties turpmākajā mārketinga analīzē.

**Visspēcīgākais slaids:**
Visspēcīgākais būs 5. slaids Rezultāti un interpretācija. Tas ne tikai parāda sausus ciparus, bet caur Feature Importance tieši un vizuāli atbild uz biznesa jautājumu – "Kas patiesībā dzen filmu ieņēmumus?". Šis slaids rosina diskusiju. Būs grafiki.

**Rīcība nezināmu jautājumu gadījumā:**
Ja auditorija uzdos jautājumu, uz kuru nezinu atbildi (piemēram, "Kā modelis uzvestos ar pilnīgi jaunu straumēšanas platformu žanru?"), es godīgi atzīšu, ka esošajā pētījumā tas netika apskatīts. Tomēr es uzreiz piedāvāšu ideju, kā es to risinātu nākotnē, piemēram, sakot: "Šobrīd man nav precīzu datu par straumēšanu, bet nākamajā iterācijā es noteikti mēģinātu pievienot platformas indikatoru kā jaunu pazīmi un izmērītu tā ietekmi."

### 9.3 Prezentācijas links

https://docs.google.com/presentation/d/1_HufosS0GIvbcPPSEaXOB3XosHAnpS_34VpxQzX5Txw/edit?usp=sharing